---
title: "Accessibility Preprocessing"
author: "Marco Stock"
---

## Setup
```{r include = FALSE}
library(here)
knitr::opts_knit$set(root.dir = here::here())
```
```{r, message = FALSE}
source("scripts/utils.R")
source("scripts/settings.R")
```

## Load genome and fasta
```{r message=FALSE, warning=FALSE}
Xenla10.1_TxDb <- loadDb("data/raw/genome/Xenla10.1.sqlite")
allgenes <- genes(Xenla10.1_TxDb)
rm(Xenla10.1_TxDb)

Xenla10.1_fa <- readDNAStringSet("data/raw/genome/XENLA_10.1_genome.fa", "fasta")
chr <- names(Xenla10.1_fa[1:18, ])
rm(Xenla10.1_fa)

allgenes <- allgenes[seqnames(allgenes) %in% chr]
promoters_chr <- promoters(allgenes, upstream = 0, downstream = 1)
```

## Load coverage data
```{r}
accessibility_dir <- 'data/raw/accessibility/'
#read Bam files i.e. coverage files as GRanges objects
pre_zga_cov1 <- readBamFileAsGRanges(bamfile = paste0(accessibility_dir, "Dam_1_SLX-20130_AR008_s_1_r.sorted.bam"), bamindex = paste0(accessibility_dir, "Dam_1_SLX-20130_AR008_s_1_r.sorted.bam.bai"))
pre_zga_cov2 <- readBamFileAsGRanges(bamfile = paste0(accessibility_dir, "Dam_1_SLX-20130_AR009_s_1_r.sorted.bam"), bamindex = paste0(accessibility_dir, "Dam_1_SLX-20130_AR009_s_1_r.sorted.bam.bai"))
pre_zga_cov <- c(pre_zga_cov1, pre_zga_cov2)

#Delete individual replicates
rm(pre_zga_cov1, pre_zga_cov2)

suppressWarnings({pre_zga_cov <- resize(pre_zga_cov, 200)})
pre_zga_cov <- trim(pre_zga_cov)
pre_zga_cov <- pre_zga_cov[seqnames(pre_zga_cov) %in% chr]
```

## Load peak calling data
```{r}
excols <- c(SignalValue = "numeric", pValue = "numeric", qValue = "numeric")

Dam1peak <- import.bed(con = paste0(accessibility_dir, "peaks/", "Dam_1_SLX-20130_AR008_s_1_r.sorted.bam_peaks01_peaks.broadPeak"), extraCols = excols)
Dam2peak <- import.bed(con = paste0(accessibility_dir, "peaks/", "Dam_1_SLX-20130_AR009_s_1_r.sorted.bam_peaks01_peaks.broadPeak"), extraCols = excols)
accessibility_peaks <- c(Dam1peak, Dam2peak)

rm(Dam1peak, Dam2peak)
strand(accessibility_peaks) <- NA
```

```{r}
accessibility_norm_mats <- list( "Pre-ZGA"=enrichment(pre_zga_cov, "Accessibility Pre-ZGA", value_name=expression("Accessibility [log"[2]*"]")))
accessibility_stages <- c("Pre-ZGA" = accessibility_peaks)
```


```{r}
promoters_1kb <- promoters(allgenes, upstream = 1000, downstream = 1000)
stages = names(accessibility_stages)
names(stages) <- stages

overlaps <- lapply(accessibility_stages, function(x) {
  findOverlaps(promoters_1kb, x, select = "first")
})

peak_widths <- lapply(stages, function(y) {
  sapply(overlaps[[y]], function(x) {
    if (is.na(x)) {
      0
    } else {
      width(accessibility_stages[[y]][x])
    }
  })
})

peak_binary <- lapply(peak_widths, function(x) {
  x > 0
})

build_df <- function(norm_matrix, binarized, widths) {
  df <- data.frame(
    "promoter.enrich" = enriched_score(norm_matrix),
    "peak" = binarized,
    "peak.width" = widths,
    "promoter.max" = apply(norm_matrix, 1, max),
    "promoter.mean" = rowMeans(norm_matrix),
    row.names = allgenes$gene_id
  )
}

accessibility_dfs <- lapply(stages, function(x) {
  build_df(accessibility_norm_mats[[x]], peak_binary[[x]], peak_widths[[x]])
})
names(accessibility_dfs) <- stages
```

```{r}
# Save data
save(accessibility_norm_mats, accessibility_dfs, accessibility_stages, file = paste0("data/processed/accessibility.RData"))
```