---
title: "DNAme Preprocessing"
author: "Marco Stock"
---

## Setup
```{r include = FALSE}
library(here)
knitr::opts_knit$set(root.dir = here::here())
```
```{r, message = FALSE}
source("scripts/utils.R")
source("scripts/settings.R")
```

## Load genome and fasta
```{r message=FALSE, warning=FALSE}
Xenla10.1_TxDb <- loadDb("data/raw/genome/Xenla10.1.sqlite")
allgenes <- genes(Xenla10.1_TxDb)
rm(Xenla10.1_TxDb)

Xenla10.1_fa <- readDNAStringSet("data/raw/genome/XENLA_10.1_genome.fa", "fasta")
chr <- names(Xenla10.1_fa[1:18, ])
rm(Xenla10.1_fa)

allgenes <- allgenes[seqnames(allgenes) %in% chr]
promoters_chr <- promoters(allgenes, upstream = 0, downstream = 1)
```

## Load DNAme Spermatid
```{r message=FALSE, warning=FALSE}
dname_spermatid_dir <- 'data/raw/dname/spermatid/'

#IP coverage replicates
spermatid_cov1 <- as(readRDS(paste0(dname_spermatid_dir, "GSM1944470_coverage.rds")), "GRanges")
spermatid_cov2 <- as(readRDS(paste0(dname_spermatid_dir, "GSM1944471_coverage.rds")), "GRanges")
spermatid_cov3 <- as(readRDS(paste0(dname_spermatid_dir, "GSM1944472_coverage.rds")), "GRanges")
spermatid_cov <- c(spermatid_cov1, spermatid_cov2, spermatid_cov3)

#Delete individual replicates
rm(spermatid_cov1, spermatid_cov2, spermatid_cov3)

DNAme_spermatid <- enrichment(spermatid_cov, "DNAme Spermatid", mode="coverage", value_name="DNAme (log2+1)", enriched_score=TRUE)
```

## Load DNAme Sperm
```{r message=FALSE, warning=FALSE}
dname_sperm_dir <- 'data/raw/dname/sperm/'

#IP coverage replicates
D2_cov1 <- readRDS(paste0(dname_sperm_dir, "GSM1944464_ranges.rds"))
D2_cov2 <- readRDS(paste0(dname_sperm_dir, "GSM1944465_ranges.rds"))
D2_cov3 <- readRDS(paste0(dname_sperm_dir, "GSM1944466_ranges.rds"))
D2_spDNAme_cov <- c(D2_cov1, D2_cov2, D2_cov3)

#Delete individual replicates
rm(D2_cov1, D2_cov2, D2_cov3)

suppressWarnings({D2_spDNAme_cov <- resize(D2_spDNAme_cov, 200)})
D2_spDNAme_cov <- trim(D2_spDNAme_cov)
D2_spDNAme_cov <- D2_spDNAme_cov[seqnames(D2_spDNAme_cov) %in% chr]
DNAme_sperm <- enrichment(D2_spDNAme_cov, "DNAme Sperm", value_name="DNAme (log2+1)", enriched_score=TRUE)
```

## Load DNAme Pre-ZGA
```{r message=FALSE, warning=FALSE}
dname_pre_zga_dir <- 'data/raw/dname/pre_zga/'

#IP coverage replicates
D4_cov1 <- readRDS(paste0(dname_pre_zga_dir, "3a_256c_IP_ranges.rds"))
D4_cov2 <- readRDS(paste0(dname_pre_zga_dir, "3b_256c_IP_ranges.rds"))
D4_S7DNAme_cov <- c(D4_cov1, D4_cov2)

#Delete individual replicates
rm(D4_cov1, D4_cov2)

D4_S7DNAme_cov <- D4_S7DNAme_cov[seqnames(D4_S7DNAme_cov) %in% chr]
DNAme_prezga <- enrichment(D4_S7DNAme_cov, "DNAme Pre-ZGA", value_name="DNAme (log2+1)", enriched_score=TRUE)
```

```{r}
# Save data
save(DNAme_spermatid, DNAme_sperm, DNAme_prezga, file = paste0("data/processed/dname.RData"))
```