---
title: "Meghana H3K4me3 Maintenance - 01 H3K4me3 analysis"
output: html_notebook
---

## Settings
```{r}
sperm_data = "Oikawa"
```


##Load genome and fasta
```{r message=FALSE, warning=FALSE}
suppressPackageStartupMessages(library(AnnotationDbi))
Xenla10.1_TxDb <- loadDb("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/NGS_data_sorted/Genome_Xl10.1/Xenla10.1.sqlite")
allgenes <- genes(Xenla10.1_TxDb)
rm(Xenla10.1_TxDb)

suppressPackageStartupMessages(library(Biostrings))
Xenla10.1_fa <- readDNAStringSet("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/NGS_data_sorted/Genome_Xl10.1/XENLA_10.1_genome.fa", "fasta")
chr <- names(Xenla10.1_fa[1:18, ])
rm(Xenla10.1_fa)

allgenes <- allgenes[seqnames(allgenes) %in% chr]
promoters_chr <- promoters(allgenes, upstream = 0, downstream = 1)
stages <- list("Spermatid", "Sperm", "Pre-ZGA", "ZGA")
```

##Utility functions
```{r}
suppressPackageStartupMessages(library(EnrichedHeatmap))
readGranges <- function(filename){
  granges <- readRDS(filename)
  suppressWarnings({granges <- resize(granges, 200)})
  granges <- trim(granges)
  granges_chr <- granges[seqnames(granges) %in% chr]
}
enrichment <- function(coverage, title){
  coverage <- normalizeToMatrix(coverage, promoters_chr, extend = 1000, mean_mode = "coverage")#, value_column = "score")
  coverage <- log2(coverage+1)
  hm <- EnrichedHeatmap(coverage, 
                        name = "H3K4me3 enrichment",
                        column_title = title,
                        col = c("blue", "red"),
                        axis_name = c("-1kb", "TSS", "+1kb"))
  draw(hm)
  return(coverage)
}
```

##Load dataset: Spermatid H3K4me3 Teperek
```{r message=FALSE, warning=FALSE}
setwd("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/Realigned_published/H3K4me3_teperek/")

#Load data
H3K4me3_spermatid.cov1 <- readGranges("GSM1944497_ranges.rds")
H3K4me3_spermatid.cov2 <- readGranges("GSM1944498_ranges.rds")
H3K4me3_spermatid.cov3 <- readGranges("GSM1944499_ranges.rds")
H3K4me3_spermatid.cov <- c(H3K4me3_spermatid.cov1, H3K4me3_spermatid.cov2, H3K4me3_spermatid.cov3)

#Merged data
H3K4me3_spermatid_raw <- enrichment(H3K4me3_spermatid.cov, "Spermatid Merged Reps")

#Individual replicates (optional)
#H3K4me3_spermatid.rep1 <- enrichment(H3K4me3_spermatid.cov1, "Spermatid Rep1")
#H3K4me3_spermatid.rep2 <- enrichment(H3K4me3_spermatid.cov2, "Spermatid Rep2")
#H3K4me3_spermatid.rep3 <- enrichment(H3K4me3_spermatid.cov3, "Spermatid Rep3")

rm(H3K4me3_spermatid.cov1, H3K4me3_spermatid.cov2, H3K4me3_spermatid.cov3, H3K4me3_spermatid.cov)
print("Dataset: Spermatid H3K4me3 Teperek loaded")
```

##Load dataset: Sperm H3K4me3 Oikawa OR Teperek
```{r warning=FALSE}
if(sperm_data == "Oikawa"){
  setwd("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/NGS_data_sorted/1_Sperm_H3K4me3_Oikawa_Xl10.1")

  #H3K4me3_sperm.cov1 <- readGranges("GSM3587294_ranges.rds")
  H3K4me3_sperm.cov2 <- readGranges("GSM3587295_ranges.rds")
  H3K4me3_sperm.cov3 <- readGranges("GSM3587296_ranges.rds")
  H3K4me3_sperm.cov <- c(H3K4me3_sperm.cov2, H3K4me3_sperm.cov3)
} else if(sperm_data == "Teperek"){
  setwd("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/Realigned_published/H3K4me3_teperek/")

  #H3K4me3_sperm.cov1 <- readGranges("GSM1944479_ranges.rds")
  H3K4me3_sperm.cov2 <- readGranges("GSM1944481_ranges.rds")
  H3K4me3_sperm.cov3 <- readGranges("GSM1944483_ranges.rds")
  H3K4me3_sperm.cov <- c(H3K4me3_sperm.cov2, H3K4me3_sperm.cov3)
}

#Merged data
H3K4me3_sperm_raw <- enrichment(H3K4me3_sperm.cov, "Sperm Merged Reps")

#Individual replicates (optional)
#H3K4me3_sperm.rep1 <- enrichment(H3K4me3_sperm.cov1, "Sperm Rep1")
#H3K4me3_sperm.rep2 <- enrichment(H3K4me3_sperm.cov2, "Sperm Rep2")
#H3K4me3_sperm.rep3 <- enrichment(H3K4me3_sperm.cov3, "Sperm Rep3")

rm(H3K4me3_sperm.cov2, H3K4me3_sperm.cov3, H3K4me3_sperm.cov)
print(paste0("Dataset: Sperm H3K4me3 ", sperm_data, " loaded"))
```


##Load dataset: Pre-ZGA (256c) H3K4me3 Eva/Oikawa
```{r warning=FALSE}
setwd("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/NGS_data_sorted/3_256c_H3K4me3_Eva-Oikawa_Xl10.1")

#IP coverage replicates
H3K4me3_prezga.cov1 <- readRDS("GSM3671368_ranges.rds")
H3K4me3_prezga.cov2 <- readRDS("GSM3671369_ranges.rds")
H3K4me3_prezga.cov <- c(H3K4me3_prezga.cov1, H3K4me3_prezga.cov2)

#Merged data
H3K4me3_prezga_raw <- enrichment(H3K4me3_prezga.cov, "256c Merged Reps")

#Individual replicates (optional)
#H3K4me3_prezga.rep1 <- enrichment(H3K4me3_prezga.cov1, "256c Rep1")
#H3K4me3_prezga.rep2 <- enrichment(H3K4me3_prezga.cov2, "256c Rep2")

rm(H3K4me3_prezga.cov1, H3K4me3_prezga.cov2, H3K4me3_prezga.cov)
print("Pre-ZGA (256c) H3K4me3 Eva/Oikawa loaded")
```

##Load dataset: ZGA (St10.5) H3K4me3 Session
```{r warning=FALSE}
setwd("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/NGS_data_sorted/6_St10.5_H3K4me3_Session_Xl10.1/GSE76059")

#IP coverage replicates
H3K4me3_zga.cov1 <- readRDS("GSM1973491_ranges.rds")
H3K4me3_zga.cov2 <- readRDS("GSM1973492_ranges.rds")
H3K4me3_zga.cov <- c(H3K4me3_zga.cov1, H3K4me3_zga.cov2)

#Merged data
H3K4me3_zga_raw <- enrichment(H3K4me3_zga.cov, "St10.5 Merged Reps")

#Individual replicates (optional)
#H3K4me3_zga.rep1 <- enrichment(H3K4me3_zga.cov1, "St10.5 Rep1")
#H3K4me3_zga.rep2 <- enrichment(H3K4me3_zga.cov2, "St10.5 Rep2")

rm(H3K4me3_zga.cov1, H3K4me3_zga.cov2, H3K4me3_zga.cov)
print("ZGA (St10.5) H3K4me3 Session loaded")
```

```{r}
chip_norm_mats <- list(H3K4me3_spermatid_raw, H3K4me3_sperm_raw, H3K4me3_prezga_raw, H3K4me3_zga_raw)
names(chip_norm_mats) <- stages
```

## Load peak calling data (broad)
```{r message=FALSE, warning=FALSE}
library(rtracklayer)
library(ggpubr)
library(EnrichedHeatmap)
#Load bed files for spermatid, sperm, pre-ZGA, post-ZGA
setwd("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/Tobias Data/Tobias-H3K4me3_ChIP_datasets_broadpeaks/GSE75164")
D0_bed1 <- import("GSM1944497_broad.bed", format = "bed")
D0_bed2 <- import("GSM1944498_broad.bed", format = "bed")
D0_bed3 <- import("GSM1944499_broad.bed", format = "bed")
D0_spdH3K4me3_bed <- c(D0_bed1, D0_bed2, D0_bed3)
strand(D0_spdH3K4me3_bed) <- NA

if(sperm_data == "Oikawa"){
  setwd("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/Tobias Data/Tobias-H3K4me3_ChIP_datasets_broadpeaks/GSE125982")
  D1_bed2 <- import("GSM3587295_broad.bed", format = "bed")
  D1_bed3 <- import("GSM3587296_broad.bed", format = "bed")
  D1_spH3K4me3_bed <- c(D1_bed2, D1_bed3)
} else if(sperm_data == "Teperek"){
  setwd("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/Tobias Data/Tobias-H3K4me3_ChIP_datasets_broadpeaks/GSE75164")
  D1_bed2 <- import("GSM1944481_broad.bed", format = "bed")
  D1_bed3 <- import("GSM1944483_broad.bed", format = "bed")
  D1_spH3K4me3_bed <- c(D1_bed2, D1_bed3)
}
strand(D1_spH3K4me3_bed) <- NA

setwd("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/Tobias Data/Tobias-H3K4me3_ChIP_datasets_broadpeaks/GSE125982")
D3_bed1 <- import("GSM3671368_broad.bed", format = "bed")
D3_bed2 <- import("GSM3671369_broad.bed", format = "bed")
D3_S7H3K4me3_bed <- c(D3_bed1, D3_bed2)
strand(D3_S7H3K4me3_bed) <- NA

setwd("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/Tobias Data/Tobias-H3K4me3_ChIP_datasets_broadpeaks/GSE76059")
D6_bed1 <- import("GSM1973491_broad.bed", format = "bed")
D6_bed2 <- import("GSM1973492_broad.bed", format = "bed")
D6_S10H3K4me3_bed <- c(D6_bed1, D6_bed2)
strand(D6_S10H3K4me3_bed) <- NA

rm(D0_bed1, D0_bed2, D0_bed3, D1_bed2, D1_bed3, D3_bed1, D3_bed2, D6_bed1, D6_bed2)

promoters_1kb <- promoters(allgenes, upstream = 1000, downstream = 1000)
h3k4stages <- list(D0_spdH3K4me3_bed, D1_spH3K4me3_bed, D3_S7H3K4me3_bed, D6_S10H3K4me3_bed)
names(h3k4stages) <- stages

overlaps <- lapply(h3k4stages, function(x){findOverlaps(promoters_1kb, x, select = "first")})
peak_widths <- lapply(stages, function(y){sapply(overlaps[[y]], function(x){if(is.na(x)){NA}else{width(h3k4stages[[y]][x])}})})
names(peak_widths) <- stages
peak_binary <- lapply(peak_widths, function(x){!is.na(x)})

build_df <- function(norm_matrix, binarized, widths){
  df <- data.frame("promoter.enrich"=enriched_score(norm_matrix), 
                   "peak"=binarized, 
                   "peak.width"=widths, 
                   "promoter.max"=apply(norm_matrix, 1, max), 
                   "promoter.mean"=rowMeans(norm_matrix),
                   row.names = allgenes$gene_id)
}

chip_dfs <- lapply(stages, function(x){build_df(chip_norm_mats[[x]], peak_binary[[x]], peak_widths[[x]])})
names(chip_dfs) <- stages
```

```{r}
#Save data
save(chip_dfs, chip_norm_mats, allgenes, promoters_chr, stages,
     file=paste0("h3k4me3_data_sperm_", sperm_data, ".RData"))
```

# Genomic distribution of peaks
```{r}
library(ChIPpeakAnno)
library(reshape2)
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}
assignChromosomeRegion <- function (peaks.RD, exon, TSS, utr5, utr3, proximal.promoter.cutoff = c(upstream = 2000, 
                                                                                                  downstream = 100), immediate.downstream.cutoff = c(upstream = 0, 
                                                                                                                                                     downstream = 1000), nucleotideLevel = FALSE, precedence = NULL, 
                                    TxDb = NULL) 
{
  if (!is.null(TxDb)) {
    if (!inherits(TxDb, "TxDb")) 
      stop("TxDb must be an object of TxDb, \n                     try\n?TxDb\tto see more info.")
    if (!inherits(peaks.RD, c("GRanges"))) 
      stop("peaks.RD must be a GRanges object.")
    if (!all(c("upstream", "downstream") %in% names(proximal.promoter.cutoff))) {
      stop("proximal.promoter.cutoff must contain elements upstream and downstream")
    }
    if (!all(c("upstream", "downstream") %in% names(immediate.downstream.cutoff))) {
      stop("immediate.downstream.cutoff must contain elements upstream and downstream")
    }
    if (!is.null(precedence)) {
      if (!all(precedence %in% c("Exons", "Introns", "fiveUTRs", 
                                 "threeUTRs", "Promoters", "immediateDownstream"))) 
        stop("precedence must be a combination of \n                         Exons, Introns, fiveUTRs, threeUTRs, \n                         Promoters, immediateDownstream")
    }
    ignore.strand <- all(as.character(strand(peaks.RD)) == 
                           "*")
    exons <- exons(TxDb, columns = NULL)
    introns <- unique(unlist(intronsByTranscript(TxDb)))
    fiveUTRs <- unique(unlist(fiveUTRsByTranscript(TxDb)))
    threeUTRs <- unique(unlist(threeUTRsByTranscript(TxDb)))
    transcripts <- unique(transcripts(TxDb, columns = NULL))
    options(warn = -1)
    try({
      promoters <- unique(promoters(TxDb, upstream = proximal.promoter.cutoff["upstream"], 
                                    downstream = proximal.promoter.cutoff["downstream"]))
      immediateDownstream <- unique(downstreams(transcripts, 
                                                upstream = immediate.downstream.cutoff["upstream"], 
                                                downstream = immediate.downstream.cutoff["downstream"]))
      promoters <- GenomicRanges::trim(promoters)
      immediateDownstream <- GenomicRanges::trim(immediateDownstream)
    })
    microRNAs <- tryCatch(microRNAs(TxDb), error = function(e) return(NULL))
    tRNAs <- tryCatch(tRNAs(TxDb), error = function(e) return(NULL))
    options(warn = 0)
    annotation <- list(exons, introns, fiveUTRs, threeUTRs, 
                       promoters, immediateDownstream)
    if (!is.null(microRNAs)) 
      annotation <- c(annotation, microRNAs = microRNAs)
    if (!is.null(tRNAs)) 
      annotation <- c(annotation, tRNAs = tRNAs)
    annotation <- lapply(annotation, function(.anno) {
      mcols(.anno) <- NULL
      .anno
    })
    names(annotation)[1:6] <- c("Exons", "Introns", "fiveUTRs", 
                                "threeUTRs", "Promoters", "immediateDownstream")
    #peaks.RD <- formatSeqnames(peaks.RD, exons)
    peaks.RD <- unique(peaks.RD)
    annotation <- GRangesList(annotation)
    newAnno <- c(unlist(annotation))
    if (ignore.strand) {
      newAnno.rd <- newAnno
      strand(newAnno.rd) <- "*"
      newAnno.rd <- reduce(trim(newAnno.rd))
      Intergenic.Region <- gaps(newAnno.rd, end = seqlengths(TxDb))
      Intergenic.Region <- Intergenic.Region[strand(Intergenic.Region) == 
                                               "*"]
    }
    else {
      newAnno.rd <- reduce(trim(newAnno))
      Intergenic.Region <- gaps(newAnno.rd, end = seqlengths(TxDb))
      Intergenic.Region <- Intergenic.Region[strand(Intergenic.Region) != 
                                               "*"]
    }
    if (!all(seqlevels(peaks.RD) %in% seqlevels(newAnno))) {
      warning("peaks.RD has sequence levels not in TxDb.")
      sharedlevels <- intersect(seqlevels(newAnno), seqlevels(peaks.RD))
      peaks.RD <- keepSeqlevels(peaks.RD, sharedlevels, 
                                pruning.mode = "coarse")
    }
    mcols(peaks.RD) <- NULL
    if (!is.null(precedence)) {
      annotation <- annotation[unique(c(precedence, names(annotation)))]
    }
    names(Intergenic.Region) <- NULL
    annotation$Intergenic.Region <- Intergenic.Region
    anno.names <- names(annotation)
    ol.anno <- findOverlaps(peaks.RD, annotation, ignore.strand = ignore.strand)
    if (nucleotideLevel) {
      jaccardIndex <- unlist(lapply(annotation, function(.ele) {
        intersection <- intersect(.ele, peaks.RD, ignore.strand = ignore.strand)
        union <- union(.ele, peaks.RD, ignore.strand = ignore.strand)
        sum(as.numeric(width(intersection)))/sum(as.numeric(width(union)))
      }))
      jaccardIndex <- jaccardIndex[anno.names]
      names(jaccardIndex) <- anno.names
      jaccardIndex[is.na(jaccardIndex)] <- 0
      newAnno <- unlist(annotation)
      newAnno$source <- rep(names(annotation), lengths(annotation))
      newAnno.disjoin <- disjoin(newAnno, with.revmap = TRUE, 
                                 ignore.strand = ignore.strand)
      if (!is.null(precedence)) {
        revmap <- cbind(from = unlist(newAnno.disjoin$revmap), 
                        to = rep(seq_along(newAnno.disjoin), lengths(newAnno.disjoin$revmap)))
        revmap <- revmap[order(revmap[, "to"], revmap[, 
                                                      "from"]), , drop = FALSE]
        revmap <- revmap[!duplicated(revmap[, "to"]), 
                         , drop = FALSE]
        newAnno.disjoin$source <- newAnno[revmap[, "from"]]$source
      }
      else {
        revmap <- unlist(newAnno.disjoin$revmap)
        newAnno.disjoin <- rep(newAnno.disjoin, lengths(newAnno.disjoin$revmap))
        newAnno.disjoin$source <- newAnno[revmap]$source
      }
      ol.anno <- findOverlaps(peaks.RD, newAnno.disjoin, 
                              ignore.strand = ignore.strand)
      queryHits <- peaks.RD[queryHits(ol.anno)]
      subjectHits <- newAnno.disjoin[subjectHits(ol.anno)]
      totalLen <- sum(as.numeric(width(peaks.RD)))
      queryHits.list <- split(queryHits, subjectHits$source)
      lens <- unlist(lapply(queryHits.list, function(.ele) sum(as.numeric(width(unique(.ele))))))
      percentage <- 100 * lens/totalLen
    }
    else {
      ol.anno.splited <- split(queryHits(ol.anno), anno.names[subjectHits(ol.anno)])
      jaccardIndex <- unlist(lapply(anno.names, function(.name) {
        union <- length(annotation[[.name]]) + length(peaks.RD) - 
          length(unique(subjectHits(findOverlaps(peaks.RD, 
                                                 annotation[[.name]], ignore.strand = ignore.strand))))
        intersection <- length(ol.anno.splited[[.name]])
        intersection/union
      }))
      names(jaccardIndex) <- anno.names
      ol.anno <- as.data.frame(ol.anno)
      ol.anno.splited <- split(ol.anno, ol.anno[, 2])
      hasAnnoHits <- do.call(rbind, ol.anno.splited[names(ol.anno.splited) != 
                                                      as.character(length(annotation))])
      hasAnnoHits <- unique(hasAnnoHits[, 1])
      ol.anno <- ol.anno[!(ol.anno[, 2] == length(annotation) & 
                             (ol.anno[, 1] %in% hasAnnoHits)), ]
      if (!is.null(precedence)) {
        ol.anno <- ol.anno[!duplicated(ol.anno[, 1]), 
        ]
      }
      subjectHits <- anno.names[ol.anno[, 2]]
      counts <- table(subjectHits)
      percentage <- 100 * counts/length(peaks.RD)
    }
    len <- length(anno.names) - length(percentage)
    if (len > 0) {
      tobeadd <- rep(0, len)
      names(tobeadd) <- anno.names[!anno.names %in% names(percentage)]
      percentage <- c(percentage, tobeadd)
    }
    percentage <- percentage[anno.names]
    return(list(percentage = percentage, jaccard = jaccardIndex))
  }
  else {
    message("Please try to use TxDb next time. Try\n\n                    ?TxDb\tto see more info.")
    annotationList <- list(exon, TSS, utr5, utr3)
    names(annotationList) <- c("Exon", "TSS", "UTR5", "UTR3")
    status <- lapply(annotationList, function(.ele) {
      if (!inherits(.ele, "GRanges")) {
        stop("Annotation of exon, TSS, utr5, utr3 must \n                         be objects of GRanges.")
      }
    })
    if (!inherits(peaks.RD, "GRanges")) 
      stop("peaks.RD must be a GRanges object.")
    ann.peaks <- annotatePeakInBatch(peaks.RD, AnnotationData = TSS)
    ann.peaks <- ann.peaks[!is.na(ann.peaks$distancetoFeature)]
    upstream <- ann.peaks[ann.peaks$insideFeature == "upstream" | 
                            (ann.peaks$distancetoFeature < 0 & ann.peaks$insideFeature == 
                               "overlapStart" & abs(ann.peaks$distancetoFeature) > 
                               ann.peaks$shortestDistance) | ann.peaks$insideFeature == 
                            "includeFeature" | (ann.peaks$distancetoFeature >= 
                                                  0 & ann.peaks$insideFeature == "overlapStart" & ann.peaks$distancetoFeature == 
                                                  ann.peaks$shortestDistance)]
    proximal.promoter.n <- length(upstream[upstream$distancetoFeature >= 
                                             -proximal.promoter.cutoff | upstream$shortestDistance <= 
                                             proximal.promoter.cutoff])
    enhancer.n <- length(upstream) - proximal.promoter.n
    downstream <- ann.peaks[ann.peaks$insideFeature == "downstream"]
    immediateDownstream.n <- length(downstream[downstream$distancetoFeature <= 
                                                 immediate.downstream.cutoff, ])
    enhancer.n <- enhancer.n + dim(downstream[downstream$distancetoFeature > 
                                                immediate.downstream.cutoff, ])
    inside.peaks <- ann.peaks[ann.peaks$insideFeature == 
                                "inside" | ann.peaks$insideFeature == "overlapEnd" | 
                                (ann.peaks$insideFeature == "overlapStart" & ann.peaks$distancetoFeature >= 
                                   0 & ann.peaks$distancetoFeature != ann.peaks$shortestDistance) | 
                                (ann.peaks$insideFeature == "overlapStart" & ann.peaks$distancetoFeature < 
                                   0 & abs(ann.peaks$distancetoFeature) == ann.peaks$shortestDistance)]
    ann.utr5.peaks <- annotatePeakInBatch(inside.peaks, AnnotationData = utr5)
    proximal.promoter.n <- proximal.promoter.n + length(ann.utr5.peaks[ann.utr5.peaks$insideFeature == 
                                                                         "upstream"])
    utr5.n <- length(ann.utr5.peaks[ann.utr5.peaks$insideFeature %in% 
                                      c("includeFeature", "inside") | (ann.utr5.peaks$insideFeature == 
                                                                         "overlapStart" & ann.utr5.peaks$distancetoFeature >= 
                                                                         0 & ann.utr5.peaks$distancetoFeature != ann.utr5.peaks$shortestDistance) | 
                                      (ann.utr5.peaks$insideFeature == "overlapStart" & 
                                         ann.utr5.peaks$distancetoFeature < 0 & abs(ann.utr5.peaks$distancetoFeature) == 
                                         ann.utr5.peaks$shortestDistance) | (ann.utr5.peaks$insideFeature == 
                                                                               "overlapEnd" & ann.utr5.peaks$strand == "+" & abs(start(ann.utr5.peaks) - 
                                                                                                                                   ann.utr5.peaks$end_position) >= (end(ann.utr5.peaks) - 
                                                                                                                                                                      ann.utr5.peaks$end_position)) | (ann.utr5.peaks$insideFeature == 
                                                                                                                                                                                                         "overlapEnd" & ann.utr5.peaks$strand == "-" & abs(end(ann.utr5.peaks) - 
                                                                                                                                                                                                                                                             ann.utr5.peaks$start_position) >= abs(start(ann.utr5.peaks) - 
                                                                                                                                                                                                                                                                                                     ann.utr5.peaks$start_position))])
    proximal.promoter.n <- proximal.promoter.n + length(ann.utr5.peaks[(ann.utr5.peaks$insideFeature == 
                                                                          "overlapStart" & ann.utr5.peaks$distancetoFeature >= 
                                                                          0 & ann.utr5.peaks$distancetoFeature == ann.utr5.peaks$shortestDistance) | 
                                                                         (ann.utr5.peaks$insideFeature == "overlapStart" & 
                                                                            ann.utr5.peaks$distancetoFeature < 0 & abs(ann.utr5.peaks$distancetoFeature) != 
                                                                            ann.utr5.peaks$shortestDistance)])
    downstream.utr5 <- ann.utr5.peaks[ann.utr5.peaks$insideFeature == 
                                        "downstream" | (ann.utr5.peaks$insideFeature == "overlapEnd" & 
                                                          ann.utr5.peaks$strand == "+" & abs(start(ann.utr5.peaks) - 
                                                                                               ann.utr5.peaks$end_position) < (end(ann.utr5.peaks) - 
                                                                                                                                 ann.utr5.peaks$end_position)) | (ann.utr5.peaks$insideFeature == 
                                                                                                                                                                    "overlapEnd" & ann.utr5.peaks$strand == "-" & abs(end(ann.utr5.peaks) - 
                                                                                                                                                                                                                        ann.utr5.peaks$start_position) < abs(start(ann.utr5.peaks) - 
                                                                                                                                                                                                                                                               ann.utr5.peaks$start_position))]
    ann.utr3.peaks <- annotatePeakInBatch(downstream.utr5, 
                                          AnnotationData = utr3)
    utr3.n <- length(ann.utr3.peaks[ann.utr3.peaks$insideFeature %in% 
                                      c("includeFeature", "overlapStart", "overlapEnd", 
                                        "inside")])
    rest.peaks <- ann.utr3.peaks[ann.utr3.peaks$insideFeature %in% 
                                   c("downstream", "upstream")]
    ann.rest.peaks <- annotatePeakInBatch(rest.peaks, AnnotationData = exon)
    intron.n <- length(ann.rest.peaks[ann.rest.peaks$insideFeature %in% 
                                        c("downstream", "upstream")])
    exon.n <- length(ann.rest.peaks) - intron.n
    total = length(peaks.RD)/100
    list(Exons = exon.n/total, Introns = intron.n/total, 
         fiveUTRs = utr5.n/total, threeUTRs = utr3.n/total, 
         Promoters = proximal.promoter.n/total, immediate.Downstream = immediateDownstream.n/total, 
         Intergenic.Region = enhancer.n/total)
  }
}

Xenla10.1_TxDb <- loadDb("~/Documents/PhD/01_Projects/NT/Meghana_H3K4me3/NGS_data_sorted/Genome_Xl10.1/Xenla10.1.sqlite")
names(h3k4stages) <- c("Spermatid", "Sperm", "Pre-ZGA", "ZGA")
genodis <- lapply(h3k4stages, function(x) {assignChromosomeRegion(x, nucleotideLevel=FALSE, proximal.promoter.cutoff = c(upstream = 1000, downstream = 1000), precedence=c("Promoters", "immediateDownstream", "fiveUTRs", "threeUTRs", "Exons", "Introns"), 
                                                                        TxDb=Xenla10.1_TxDb)
})
b <- matrix(unlist(genodis), nrow =14)
genodis2 <- b[1:7, ]
colnames(genodis2) <- names(h3k4stages)
rownames(genodis2) <- c("Promoters", "ImmediateDownstream", "FiveUTRs", "ThreeUTRs", "Exons", "Introns", "Intergenic regions")
ggdf1 <- melt(genodis2)
ggdf1$Var1 <- factor(ggdf1$Var1, levels=rev(rownames(genodis2)))
ggplot(ggdf1, aes(fill=Var1, y=value, x=Var2)) + geom_bar(position="stack", stat="identity") + theme_bw() + ylab("%") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.title.x = element_blank(), legend.title=element_blank()) + ggtitle("H3K4me3 peaks") + scale_fill_manual(values = c("grey", gg_color_hue(6)))
rm(Xenla10.1_TxDb)
```






