---
title: "DNAme analysis"
author: "Marco Stock"
---

## Setup
```{r include = FALSE}
library(here)
knitr::opts_knit$set(root.dir = here::here())
```
```{r, message=FALSE}
source("scripts/utils.R")
source("scripts/settings.R")
```
```{r}
load("data/processed/dname.RData")
```

## Genomic distribution of peaks
```{r}
gg_color_hue <- function(n) {
  hues <- seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}

Xenla10.1_TxDb <- loadDb("data/raw/genome/Xenla10.1.sqlite")
genodis <- lapply(DNAme_stages, function(x) {
  assignChromosomeRegion(x,
    nucleotideLevel = FALSE, proximal.promoter.cutoff = c(upstream = 1000, downstream = 1000), precedence = c("Promoters", "immediateDownstream", "fiveUTRs", "threeUTRs", "Exons", "Introns"),
    TxDb = Xenla10.1_TxDb
  )
})
b <- matrix(unlist(genodis), nrow = 14)
genodis2 <- b[1:7, ]
colnames(genodis2) <- names(DNAme_stages)
rownames(genodis2) <- c("Promoters", "ImmediateDownstream", "FiveUTRs", "ThreeUTRs", "Exons", "Introns", "Intergenic regions")
ggdf1 <- melt(genodis2)
ggdf1$Var1 <- factor(ggdf1$Var1, levels = rev(rownames(genodis2)))
ggplot(ggdf1, aes(fill = Var1, y = value, x = Var2)) +
  geom_bar(position = "stack", stat = "identity") +
  theme_bw() +
  ylab("%") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.title.x = element_blank(), legend.title = element_blank()) +
  ggtitle("DNAme peaks") +
  scale_fill_manual(values = c("grey", gg_color_hue(6)))
ggsave('output/dname/genomic_distribution.pdf', width=3.3, height=3)

rm(Xenla10.1_TxDb)
```

## DNAme dynamics
```{r}
DNAme_dfs[["ZGA"]] <- DNAme_dfs[["Pre-ZGA"]]

dname_groups <- list(Kept = "Y_Y_Y_Y", Recovered=c("Y_Y_N_Y", "Y_N_Y_Y", "Y_N_N_Y"), Lost=c("Y_Y_Y_N", "Y_Y_N_N", "Y_N_N_N"), Gained=c("N_N_Y_Y", "N_N_N_Y"), Absent="N_N_N_N")
dname_names_complete <- c("Y_Y_Y_Y", "Y_Y_Y_N", "Y_Y_N_Y", "Y_Y_N_N", "Y_N_Y_Y", "Y_N_Y_N", "Y_N_N_Y", "Y_N_N_N", "N_Y_Y_Y", "N_Y_Y_N", "N_Y_N_Y", "N_Y_N_N", "N_N_Y_Y", "N_N_Y_N", "N_N_N_Y", "N_N_N_N")
dname_groups_complete <- setNames(as.list(dname_names_complete), dname_names_complete)
peak_combinations <- compute_peak_combinations(DNAme_dfs)

for (combination_name in names(peak_combinations)) {
  assign(combination_name, peak_combinations[[combination_name]])
}

get_dyns <- function(cluster){
    lists = as.vector(unlist(sapply(cluster, function(dyn_name){get(dyn_name)}, simplify = TRUE), use.names = FALSE))}

dname_dyns <- lapply(dname_groups, get_dyns)
dname_dyns_complete <- lapply(dname_groups_complete, get_dyns)
dname_dyns_sizes = unlist(lapply(dname_dyns, function(x) length(x)))
```

## Venn diagram of DNAme dynamics
```{r}
groups = dname_groups
cols = cols_dname
euler_cols <- setNames(rep("#FFFFFF00", 16), dname_names_complete)
for(i in 1:length(groups)){
  cluster <- groups[[i]]
  for(dynamic in cluster){
    euler_cols[dynamic]=cols[names(groups)[i]]
  }
}
plot(euler(c("Spermatid"=length(Y_N_N_N), "Sperm"=length(N_Y_N_N), "Pre-ZGA"=length(N_N_Y_N), "ZGA"=length(N_N_N_Y),
           "Spermatid&Sperm"=length(Y_Y_N_N), "Spermatid&Pre-ZGA"=length(Y_N_Y_N), "Spermatid&ZGA"=length(Y_N_N_Y),
           "Sperm&Pre-ZGA"=length(N_Y_Y_N), "Sperm&ZGA"=length(N_Y_N_Y), "Pre-ZGA&ZGA"=length(N_N_Y_Y),
      "Spermatid&Sperm&Pre-ZGA"=length(Y_Y_Y_N), "Spermatid&Sperm&ZGA"=length(Y_Y_N_Y), "Spermatid&Pre-ZGA&ZGA"=length(Y_N_Y_Y),
      "Sperm&Pre-ZGA&ZGA"=length(N_Y_Y_Y), "Spermatid&Sperm&Pre-ZGA&ZGA"=length(Y_Y_Y_Y))),
     quantities = TRUE,  fills=euler_cols[c("Y_N_N_N", "N_Y_N_N", "N_N_Y_N", "N_N_N_Y", "Y_Y_N_N", "Y_N_Y_N", "Y_N_N_Y", "N_Y_Y_N", "N_Y_N_Y", "N_N_Y_Y", "Y_Y_Y_N", "Y_Y_N_Y", "Y_N_Y_Y", "N_Y_Y_Y", "Y_Y_Y_Y")])

plot_venn <- function() {
  plot(venn(c("Spermatid"=length(Y_N_N_N), "Sperm"=length(N_Y_N_N), "Pre-ZGA"=length(N_N_Y_N), "ZGA"=length(N_N_N_Y),
             "Spermatid&Sperm"=length(Y_Y_N_N), "Spermatid&Pre-ZGA"=length(Y_N_Y_N), "Spermatid&ZGA"=length(Y_N_N_Y),
             "Sperm&Pre-ZGA"=length(N_Y_Y_N), "Sperm&ZGA"=length(N_Y_N_Y), "Pre-ZGA&ZGA"=length(N_N_Y_Y),
        "Spermatid&Sperm&Pre-ZGA"=length(Y_Y_Y_N), "Spermatid&Sperm&ZGA"=length(Y_Y_N_Y), "Spermatid&Pre-ZGA&ZGA"=length(Y_N_Y_Y),
        "Sperm&Pre-ZGA&ZGA"=length(N_Y_Y_Y), "Spermatid&Sperm&Pre-ZGA&ZGA"=length(Y_Y_Y_Y))),
       quantities = TRUE,  fills=euler_cols[c("Y_N_N_N", "N_Y_N_N", "N_N_Y_N", "N_N_N_Y", "Y_Y_N_N", "Y_N_Y_N", "Y_N_N_Y", "N_Y_Y_N", "N_Y_N_Y", "N_N_Y_Y", "Y_Y_Y_N", "Y_Y_N_Y", "Y_N_Y_Y", "N_Y_Y_Y", "Y_Y_Y_Y")])
}

pdf(file="output/dname/dname_dynamics_venn.pdf", width=5, height=5)
plot_venn()
dev.off()

plot_venn()
```
```{r fig.height=5, fig.width=2}
dyns_sizes = dname_dyns_sizes
cols = cols_dname
ggplot(data.frame("sizes"=dyns_sizes, "dyn"=names(dyns_sizes), "id"=1), aes(fill=dyn, y=sizes, x=id)) + 
    geom_bar(position="fill", stat="identity", fill=cols) + theme_pubr() + ylab("Ratio of genome") + theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
ggsave('output/dname/dname_dynamics_stacked_bar.pdf', width=2, height=5)
```
## Alluvial plots
```{r message=FALSE, warning=FALSE}
dyns = dna_me_dyns

plot_alluvial <- function(remove_NNNN, remove_YYYY, equal_scale){
  get_row <- function(dyns){
    levels = unlist(strsplit(names(dyns), split = "_"))
    return(c(levels, length(dyns[[1]]), names(dyns)))
  }
  plot_df = data.frame()
  for (i in 1:length(dname_dyns_complete)){
    row = get_row(dname_dyns_complete[i])
    plot_df = rbind(plot_df, row)
  }
  
  colnames(plot_df) <- c(stages, "Freq", "Dynamic")
  plot_df <- plot_df %>% mutate_at("Freq", as.numeric)
  
  if(remove_NNNN){plot_df <- plot_df[!plot_df$Dynamic %in% c("N_N_N_N"),]}
  if(remove_YYYY){plot_df <- plot_df[!plot_df$Dynamic %in% c("Y_Y_Y_Y"),]}
  
  plot_df <- plot_df[!plot_df$Dynamic %in% c("Y_N_Y_N", "N_Y_N_Y"),]
  
  for(stage in stages){
    plot_df[,stage] <- factor(plot_df[,stage], levels=c("Y", "N"))
  }
  if(equal_scale){
    plot_df$Freq <- 1
  }else{
    plot_df <- plot_df[!plot_df$Dynamic %in% c("N_Y_N_N", "N_Y_N_Y", "N_Y_Y_Y", "N_Y_Y_N", "Y_N_Y_N", "N_N_Y_N"),]
  }
  alluvial_plot <- ggplot(plot_df,
         aes(y = Freq, axis1 = Spermatid, axis2 = Sperm, axis3= `Pre-ZGA`, axis4 = ZGA, fill=Dynamic)) +
    geom_alluvium(aes(fill = Dynamic), width = 1/12, lode.guidance="backfront") + theme_pubr() +
    geom_stratum(width = 1/12, alpha = .75) +
    geom_label(stat = "stratum", aes(label = after_stat(stratum)), fill="white") +
    scale_x_discrete(limits = c("Spermatid", "Sperm", "Pre-ZGA", "ZGA"), expand = c(.05, .05)) +
    scale_fill_manual(values=euler_cols) + 
    guides(fill="none") 
  if(equal_scale){
    alluvial_plot <- alluvial_plot + geom_hline(yintercept = 1/2*sum(plot_df$Freq), col = "black", linetype="dotted") +
    theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())
  }
  plot(alluvial_plot)
}

pdf(file="output/dname/dname_dynamics_alluvial_full.pdf", width=5, height=3)
plot_alluvial(remove_NNNN = FALSE, remove_YYYY = FALSE, equal_scale=FALSE)
dev.off()

pdf(file="output/dname/dname_dynamics_alluvial_variable.pdf", width=5, height=3)
plot_alluvial(remove_NNNN = TRUE, remove_YYYY = TRUE, equal_scale=FALSE)
dev.off()

plot_alluvial(remove_NNNN = FALSE, remove_YYYY = FALSE, equal_scale=FALSE)
plot_alluvial(remove_NNNN = TRUE, remove_YYYY = TRUE, equal_scale=FALSE)

```
